allennlp/training/trainer.py

```diff
@@ -716,22 +711,15 @@ def train(self) -> Dict[str, Any]:
 
         return metrics
 
-    def _save_checkpoint(self, epoch: Union[int, str]) -> None:
-        """
-        Saves a checkpoint of the model to self._serialization_dir.
-        Is a no-op if self._serialization_dir is None.
-
-        # Parameters
-
-        epoch : Union[int, str], required.
-            The epoch of training.  If the checkpoint is saved in the middle
-            of an epoch, the parameter is a string with the epoch and timestamp.
-        """
-        # If moving averages are used for parameters, we save
-        # the moving average values into checkpoint, instead of the current values.
+    @overrides
+    def prep_state_for_checkpointing(self) -> Tuple[Dict[str, Any], Dict[str, Any]]:
         if self._moving_average is not None:
+            # Assigning average value to model parameters.  The checkpointer will call
+            # `restore_state_after_checkpointing` when it is done to put this back to what it was.
```

Comment:
Should this be a context manager then?

Reply:
I'm not sure how to actually do that.  Another wrinkle in here is that the checkpointer might decide to decline actually saving anything when the `save_checkpoint` method is called, and we don't want to do extra copies in that case.  Maybe you're talking about having something like this in the checkpointer?

```python
def save_checkpoint(..., trainer, ...):
    with trainer.get_checkpoint_state():
        # save stuff
    # all of the cleanup logic that's there
```

Yes, I think this could work, and I like it much better than what I did.  I've never actually written a context manager before, so I'll have to figure out how to do this, but it looks like it should work.

Reply:
Done, and yeah, I like this a whole lot better.  Thanks for the suggestion!